snap/squashfs2/squashfs2.go

```diff
@@ -0,0 +1,531 @@
+}
+
+func readInt32(data []byte) int32 {
+	return int32(readUint32(data))
+}
+
+func readUint64(data []byte) uint64 {
+	return uint64(data[0]) | uint64(data[1])<<8 | uint64(data[2])<<16 | uint64(data[3])<<24 |
+		uint64(data[4])<<32 | uint64(data[5])<<40 | uint64(data[6])<<48 | uint64(data[7])<<56
+}
+
+func readInt64(data []byte) int64 {
+	return int64(readUint64(data))
+}
+
+func readInodeRef(data []byte) metadataRef {
+	value := readInt64(data)
+	return metadataRef{
+		offset: int(value & 0xffff),
+		block:  value >> 16,
+	}
+}
+
+func parseSuperBlock(data []byte) (*squashfs_superblock, error) {
+	if len(data) < superblockSize {
+		return nil, fmt.Errorf("squashfs: superblock too small")
+	}
+
+	sb := &squashfs_superblock{}
+	copy(sb.magic[:], data[:4])
+	if sb.magic != magic {
+		return nil, fmt.Errorf("squashfs: invalid magic")
+	}
+
+	sb.inodes = readUint32(data[4:])
+	sb.mkfsTime = readUint32(data[8:])
+	sb.blockSize = readUint32(data[12:])
+	sb.fragments = readUint32(data[16:])
+	sb.compressionType = readUint16(data[20:])
+	sb.blockSizeLog2 = readUint16(data[22:])
+	sb.flags = readUint16(data[24:])
+	sb.noIDs = readUint16(data[26:])
+	sb.s_major = readUint16(data[28:])
+	sb.s_minor = readUint16(data[30:])
+	sb.rootIno = readInodeRef(data[32:])
+	sb.bytesUsed = readInt64(data[40:])
+	sb.idTableStart = readInt64(data[48:])
+	sb.xattrIdTableSz = readInt64(data[56:])
+	sb.inodeTable = readInt64(data[64:])
+	sb.directoryTable = readInt64(data[72:])
+	sb.fragmentTable = readInt64(data[80:])
+	sb.lookupTable = readInt64(data[88:])
+	return sb, nil
+}
+
+func readSuperBlock(stream *os.File) (*squashfs_superblock, error) {
+	buffer := make([]byte, superblockSize)
+	_, err := stream.Read(buffer)
+	if err != nil {
+		return nil, err
+	}
+	return parseSuperBlock(buffer)
+}
+
+func createCompressionBackend(stream *os.File, sb *squashfs_superblock) (CompressionBackend, error) {
+	println("squashfs: compression type", sb.compressionType)
+	var optionsBlock *metaBlockReader = nil
+	if sb.flags&superBlockCompressorOptions != 0 {
+		optionsBlock = metablockReaderCreate(stream, nil, superblockSize)
+	}
+
+	switch sb.compressionType {
+	case compressionXz:
+		return createXzBackend(optionsBlock)
+	default:
+		return nil, fmt.Errorf("squashfs: unsupported compression type %d", sb.compressionType)
+	}
+}
+
+// SquashFS layout
+// ---------------
+// |  superblock   |
+// |---------------|
+// |  compression  |
+// |    options    |
+// |---------------|
+// |  datablocks   |
+// |  & fragments  |
+// |---------------|
+// |  inode table  |
+// |---------------|
+// |   directory   |
+// |     table     |
+// |---------------|
+// |   fragment    |
+// |    table      |
+// |---------------|
+// |    export     |
+// |    table      |
+// |---------------|
+// |    uid/gid    |
+// |  lookup table |
+// |---------------|
+// |     xattr     |
+// |     table     |
+// |---------------|
+func SquashFsOpen(path string) (*SquashFileSystem, error) {
```

Comment:
Given that the package is already called `squashfs` when using this it would read `squashfs.SquashFsOpen()` which is a bit redundant. Maybe just: `squashfs.Open()` or `squashfs.FsOpen()`?